eng/Microsoft.DotNet.Arcade.Sdk/tools/git-clone-to-dir.sh

```diff
@@ -0,0 +1,107 @@
+#!/usr/bin/env bash
+
+usage() {
+  echo "$0 [options]"
+  echo "Creates a git clone of <source> into <dest> directory, optionally saving WIP changes."
+  echo ""
+  echo "Required:"
+  echo "  --source <dir>  (Required) The source Git directory."
+  echo "  --dest <dir>    (Required) The destination Git directory. Created if doesn't exist."
+  echo ""
+  echo "Optional:"
+  echo "  --clean     If dest directory already exists, delete it."
+  echo "  --copy-wip  Transfer most types of uncommitted change into the"
+  echo "              destination directory. Useful for dev workflows."
+  echo "  -h --help   Print help and exit."
+}
+
+# An alternative to "git clone" would be to use a "git worktree". Potential benefits:
+# * This has more integration with the user's normal repo. If they make ad-hoc changes in the
+#   worktree, it is easy to cherry-pick onto the developer's "real" branch.
+# * A worktree uses a '.git' file rather than a full '.git' directory, which might have storage
+#   implications. (However, a local clone's '.git' directory uses hard links to save space/time,
+#   so it's not certain that this affects performance at all.)
+# 
+# Downside of worktrees:
+# * Some configuration is set up in '.git/worktrees' which may be difficult to coordinate
+#   properly. In particular, if the user is already using worktrees for their own purposes, we
+#   would have to be careful that running source-build in one worktree doesn't interfere with
+#   source-build in the other worktree.
+
+set -euo pipefail
+
+while [[ $# > 0 ]]; do
+  opt="$(echo "$1" | awk '{print tolower($0)}')"
+  case "$opt" in
+    -h|--help)
+      usage
+      exit 0
+      ;;
+    --source)
+      sourceDir="$2"
+      shift
+      ;;
+    --dest)
+      destDir="$2"
+      shift
+      ;;
+    --clean)
+      clean=1
+      ;;
+    --copy-wip)
+      copyWip=1
+      ;;
+    *)
+      echo "Unrecognized option: $1"
+      echo ""
+      usage
+      exit 1
+      ;;
+  esac
+  shift
+done
+
+[ ! -d "${sourceDir:-}" ] && echo "--source not a directory: $sourceDir" && exit 1
+
+[ ! "${destDir:-}" ] && echo "--dest not specified" && exit 1
+
+if [ -e "$destDir" ]; then
+  echo "Destination already exists!"
+  if [ -f "$destDir" ]; then
+    echo "Existing destination is a regular file, not a directory. This is unusual: aborting."
+    exit 1
+  elif [ "${clean:-}" ]; then
+    echo "Clean is enabled: removing $destDir and continuing..."
+    rm -rf "$destDir"
+  else
+    echo "Clean is not enabled: aborting."
+    exit 1
+  fi
+fi
+
+if [ ! -e "$destDir" ]; then
+  mkdir -p "$destDir"
+
+  if [ "${copyWip:-}" ]; then
+    echo "WIP copying is enabled"
+    # Copy over changes that haven't been committed, for dev inner loop.
+    # This gets changes (whether staged or not) but misses untracked files.
```

Comment:
Can add `-u` if you want to catch untracked files.

Reply:
Unfortunately that doesn't work... I had it in there a while thinking it was doing something, but I realized that was just giving the stash a name:

```
## dagood ~/temp/u (master)
#> git status
On branch master
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        modified:   a

Untracked files:
  (use "git add <file>..." to include in what will be committed)
        b

no changes added to commit (use "git add" and/or "git commit -a")
## dagood ~/temp/u (master)
#> git stash create -u
5953821594fab912900019914355564452cd5b41
## dagood ~/temp/u (master)
#> git log 5953821594fab912900019914355564452cd5b41
commit 5953821594fab912900019914355564452cd5b41
Merge: 37b9ea8 4152657
Author: Davis Goodin <dagood@microsoft.com>
Date:   Wed Aug 12 15:30:11 2020 -0500

    On master: -u

commit 4152657a9d95689ab7a6dbd6ccb3819abea5ad44
Author: Davis Goodin <dagood@microsoft.com>
Date:   Wed Aug 12 15:30:11 2020 -0500

    index on master: 37b9ea8 Initial

commit 37b9ea81e5188862dae2a8d70ddf7be7931e517b (HEAD -> master)
Author: Davis Goodin <dagood@microsoft.com>
Date:   Wed Aug 12 15:29:41 2020 -0500

    Initial
## dagood ~/temp/u (master)
#> git show 4152657a9d95689ab7a6dbd6ccb3819abea5ad44
commit 4152657a9d95689ab7a6dbd6ccb3819abea5ad44
Author: Davis Goodin <dagood@microsoft.com>
Date:   Wed Aug 12 15:30:11 2020 -0500

    index on master: 37b9ea8 Initial
## dagood ~/temp/u (master)
#> git show 5953821594fab912900019914355564452cd5b41
commit 5953821594fab912900019914355564452cd5b41
Merge: 37b9ea8 4152657
Author: Davis Goodin <dagood@microsoft.com>
Date:   Wed Aug 12 15:30:11 2020 -0500

    On master: -u

diff --cc a
index 7898192,7898192..f2ad6c7
--- a/a
+++ b/a
@@@ -1,1 -1,1 +1,1 @@@
--a
++c
```

Reply:
Well that's weird.  Sorry about that, thanks for the info!